jocd/src/br/org/certi/jocd/dapaccess/connectioninterface/AndroidUsbDevice.java

```diff
@@ -0,0 +1,467 @@
+                int index = this.interfaceNumber;
+                deviceConnection.controlTransfer(requestType, request, value, index, packet,
+                        data.length, timeout);
+                return;
+            }
+
+            // If we got here, means that we have an output endpoint.
+            int written = deviceConnection.bulkTransfer(outputEndpoint, packet, packet.length,
+                    timeout);
+        }
+    }
+
+    /*
+     * Open the device.
+     */
+    public void open() {
+        // Do it once, and break to clean if anything goes wrong.
+        do {
+            if (connection != null) {
+                Log.e(TAG, "Trying to open device while is already opened.");
+                return;
+            }
+
+            if (device == null) {
+                Log.e(TAG, "Trying to open device a null device.");
+                return;
+            }
+
+            // Check if we have permission.
+            if (!usbManager.hasPermission(device)) {
+                Log.w(TAG, appName + " doesn't have permission to access device with Product ID: "
+                        + device.getProductId() + ", Vendor ID: " + device.getVendorId() + " and " +
+                        "serial number: " + device.getSerialNumber());
+                PendingIntent pi = PendingIntent.getBroadcast(context, 0,
+                        new Intent(actionUsbPermission), 0);
+                usbManager.requestPermission(device, pi);
+                return;
+            }
+
+            // Attempt to open the device.
+            this.deviceConnection = usbManager.openDevice(device);
+            if (this.deviceConnection == null) {
+                Log.e(TAG, "Unable to open device.");
+                return;
+            }
+
+            // Look for the HID interface.
+            if (!lookForHidInterface()) {
+                Log.e(TAG, "Couldn't find any HID device.");
+                break;
+            }
+
+            // Find endpoints.
+            if (!findEndpoints()) {
+                Log.e(TAG, "Couldn't find endpoints.");
+                break;
+            }
+
+            // Claim interface.
+            if (!this.deviceConnection.claimInterface(usbInterface, true))
+            {
+                Log.e(TAG, "Couldn't claim interface.");
+                break;
+            }
+
+            // Start rx thread.
+            this.open = true;
+            this.startRxThread();
+
+            // If everything ok, leave (don't let it get to clean section below).
+            return;
+
+        } while (false);
+
+
+        // We should never get here... unless something went wrong.
+        // Clean whatever we done here.
+        this.deviceConnection = null;
+        this.inputEndpoint = null;
+        this.outputEndpoint = null;
+        this.interfaceNumber = -1;
+        this.open = false;
+    }
+
+    /*
+     * Close the device.
+     */
+    public void close() {
+        if (connection == null) return;
```

Comment:
According to the Google style guide, even where it is optional, braces must be used.